src/Extensions/Activities/Internal/EventSource/ActivityEventSender.cs

```diff
@@ -0,0 +1,128 @@
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using Microsoft.Extensions.Logging;
+using Microsoft.Omex.Extensions.Abstractions;
+using Microsoft.Omex.Extensions.Abstractions.Activities;
+using Microsoft.Omex.Extensions.Abstractions.Activities.Processing;
+using Microsoft.Omex.Extensions.Abstractions.ExecutionContext;
+
+namespace Microsoft.Omex.Extensions.Activities
+{
+	internal sealed class ActivityEventSender : IActivitiesEventSender
+	{
+		public ActivityEventSender(ActivityEventSource eventSource, IExecutionContext executionContext, ILogger<ActivityEventSender> logger)
+		{
+			m_eventSource = eventSource;
+			m_serviceName = executionContext.ServiceName;
+			m_logger = logger;
+		}
+
+		public void LogActivityStop(Activity activity)
+		{
+			if (!m_eventSource.IsEnabled())
+			{
+				return;
+			}
+
+			string serviceName = m_serviceName;
+			string activityId = activity.Id ?? string.Empty;
+			string name = activity.OperationName;
+			double durationMs = activity.Duration.TotalMilliseconds;
+			string userHash = activity.GetUserHash(); //TODO: We need add middleware that will set userhash in compliant way and IsTransaction GitHub Issue #166
+			bool isHealthCheck = activity.IsHealthCheck();
+
+			string subtype = NullPlaceholder;
+			string metadata = NullPlaceholder;
+			string resultAsString = NullPlaceholder;
+			foreach (KeyValuePair<string, string?> pair in activity.Tags)
+			{
+				if (pair.Value == null)
+				{
+					continue;
+				}
+
+				if (string.Equals(ActivityTagKeys.Result, pair.Key, StringComparison.Ordinal))
+				{
+					resultAsString = pair.Value;
+				}
+				else if (string.Equals(ActivityTagKeys.SubType, pair.Key, StringComparison.Ordinal))
+				{
+					subtype = pair.Value;
+				}
+				else if (string.Equals(ActivityTagKeys.Metadata, pair.Key, StringComparison.Ordinal))
+				{
+					metadata = pair.Value;
+				}
+			}
+
+#pragma warning disable CS0618 // Until it's used we need to include correlationId into events
+			string correlationId = activity.GetObsoleteCorrelationId()?.ToString()
+				?? activity.GetRootIdAsGuid()?.ToString()
+				?? NullPlaceholder;
+#pragma warning restore CS0618
+
+			string nameAsString = SanitizeString(name, nameof(name), name);
+			string subTypeAsString = SanitizeString(subtype, nameof(subtype), name);
+			string metaDataAsString = SanitizeString(metadata, nameof(metadata), name);
+			string userHashAsString = SanitizeString(userHash, nameof(userHash), name);
+			string serviceNameAsString = SanitizeString(serviceName, nameof(serviceName), name);
+			string correlationIdAsString = SanitizeString(correlationId, nameof(correlationId), name);
+			string activityIdAsString = SanitizeString(activityId, nameof(activityId), name);
+			long durationMsAsLong = Convert.ToInt64(durationMs, CultureInfo.InvariantCulture);
+
+			if (isHealthCheck)
+			{
+				m_eventSource.WriteTimedScopeTestEvent(
```

Comment:
This is still named timedscope?

Reply:
Thanks, I'll check if it could be renamed with influencing ETW events

Reply:
Renamed as much as possible